state/protocol/badger/mutator.go

```diff
@@ -168,34 +172,38 @@ func (m *FollowerState) ExtendCertified(ctx context.Context, candidate *flow.Blo
 		return fmt.Errorf("qc doesn't certify candidate block, expect %x blockID, got %x", blockID, certifyingQC.BlockID)
 	}
 
-	// check if the block header is a valid extension of parent block
-	err = m.headerExtend(ctx, candidate, certifyingQC, deferredDbOps)
-	if err != nil {
-		// since we have a QC for this block, it cannot be an invalid extension
-		return fmt.Errorf("unexpected invalid block (id=%x) with certifying qc (id=%x): %s",
-			candidate.ID(), certifyingQC.ID(), err.Error())
-	}
-
-	// find the last seal at the parent block
-	_, err = m.lastSealed(candidate, deferredDbOps)
-	if err != nil {
-		return fmt.Errorf("failed to determine the lastest sealed block in fork: %w", err)
-	}
+	return m.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
+		// check if the block header is a valid extension of parent block
+		err = m.headerExtend(ctx, candidate, certifyingQC, rw)
+		if err != nil {
+			// since we have a QC for this block, it cannot be an invalid extension
+			return fmt.Errorf("unexpected invalid block (id=%x) with certifying qc (id=%x): %s",
+				candidate.ID(), certifyingQC.ID(), err.Error())
+		}
 
-	// evolve protocol state and verify consistency with commitment included in
-	err = m.evolveProtocolState(ctx, candidate, deferredDbOps)
-	if err != nil {
-		return fmt.Errorf("evolving protocol state failed: %w", err)
-	}
+		// find the last seal at the parent block
+		_, err = m.lastSealed(candidate, rw)
+		if err != nil {
+			return fmt.Errorf("failed to determine the lastest sealed block in fork: %w", err)
+		}
 
-	// Execute the deferred database operations as one atomic transaction and emit scheduled notifications on success.
-	// The `candidate` block _must be valid_ (otherwise, the state will be corrupted)!
-	err = operation.RetryOnConflictTx(m.db, transaction.Update, deferredDbOps.Pending()) // No errors are expected during normal operations
-	if err != nil {
-		return fmt.Errorf("failed to persist candidate block %v and its dependencies: %w", blockID, err)
-	}
+		// evolve protocol state and verify consistency with commitment included in
+		dbUpdates, err := m.evolveProtocolState(ctx, candidate, rw)
```

Comment:
TODO (from @jordanschalm ) we might not need the deferred db updates, because the candidate passed into the `Extend` method has already been fully constructed.